File: surf-api-core/surf-api-core-api/src/main/java/dev/slne/surf/surfapi/core/api/util/BlockStateFactory.java
package dev.slne.surf.surfapi.core.api.util;

import com.github.retrooper.packetevents.protocol.world.BlockFace;
import com.github.retrooper.packetevents.protocol.world.states.WrappedBlockState;
import com.github.retrooper.packetevents.protocol.world.states.enums.*;
import com.github.retrooper.packetevents.protocol.world.states.type.StateType;
import org.jetbrains.annotations.NotNull;

import javax.annotation.ParametersAreNonnullByDefault;

import static com.google.common.base.Preconditions.*;

public interface BlockStateFactory {

    static Builder builder(@NotNull StateType stateType) {
        return new BlockStateFactoryImpl.BuilderImpl(checkNotNull(stateType, "stateType").createBlockState().clone());
    }

    static Builder builder(@NotNull WrappedBlockState wrappedBlockState) {
        return new BlockStateFactoryImpl.BuilderImpl(checkNotNull(wrappedBlockState, "wrappedBlockState").clone());
    }

    static WrappedBlockState of(@NotNull StateType stateType) {
        return checkNotNull(stateType, "stateType").createBlockState().clone();
    }

    @ParametersAreNonnullByDefault
    @SuppressWarnings("unused")
    interface Builder {
        int age();

        Builder age(int age);

        boolean attached();

        Builder attached(boolean attached);

        Attachment attachment();

        Builder attachment(@NotNull Attachment attachment);

        Axis axis();

        Builder axis(@NotNull Axis axis);

        boolean berries();

        Builder berries(boolean berries);

        int bites();

        Builder bites(int bites);

        boolean bottom();

        Builder bottom(boolean bottom);

        int candles();

        Builder candles(int candles);

        int charges();

        Builder charges(int charges);

        boolean conditional();

        Builder conditional(boolean conditional);

        int delay();

        Builder delay(int delay);

        boolean disarmed();

        Builder disarmed(boolean disarmed);

        int distance();

        Builder distance(int distance);

        boolean down();

        Builder down(boolean down);

        boolean drag();

        Builder drag(boolean drag);

        int eggs();

        Builder eggs(int eggs);

        boolean enabled();

        Builder enabled(boolean enabled);

        boolean extended();

        Builder extended(boolean extended);

        boolean eye();

        Builder eye(boolean eye);

        Face face();

        Builder face(@NotNull Face face);

        BlockFace facing();

        Builder facing(@NotNull BlockFace facing);

        Half half();

        Builder half(@NotNull Half half);

        boolean hanging();

        Builder hanging(boolean hanging);

        boolean hasBook();

        Builder hasBook(boolean hasBook);

        boolean hasBottle0();

        Builder hasBottle0(boolean hasBottle0);

        boolean hasBottle1();

        Builder hasBottle1(boolean hasBottle1);

        boolean hasBottle2();

        Builder hasBottle2(boolean hasBottle2);

        boolean hasRecord();

        Builder hasRecord(boolean hasRecord);

        int hatch();

        Builder hatch(int hatch);

        Hinge hinge();

        Builder hinge(@NotNull Hinge hinge);

        int honeyLevel();

        Builder honeyLevel(int honeyLevel);

        boolean inWall();

        Builder inWall(boolean inWall);

        Instrument instrument();

        Builder instrument(@NotNull Instrument instrument);

        boolean inverted();

        Builder inverted(boolean inverted);

        int layers();

        Builder layers(int layers);

        Leaves leaves();

        Builder leaves(Leaves leaves);

        int level();

        Builder level(int level);

        boolean lit();

        Builder lit(boolean lit);

        boolean locked();

        Builder locked(boolean locked);

        Mode mode();

        Builder mode(@NotNull Mode mode);

        int moisture();

        Builder moisture(int moisture);

        North north();

        Builder north(@NotNull North north);

        int note();

        Builder note(int note);

        boolean occupied();

        Builder occupied(boolean occupied);

        boolean shrieking();

        Builder shrieking(boolean shrieking);

        boolean canSummon();

        Builder canSummon(boolean canSummon);

        boolean open();

        Builder open(boolean open);

        Orientation orientation();

        Builder orientation(Orientation orientation);

        Part part();

        Builder part(Part part);

        boolean persistent();

        Builder persistent(boolean persistent);

        int pickles();

        Builder pickles(int pickles);

        int power();

        Builder power(int power);

        boolean powered();

        Builder powered(boolean powered);

        int rotation();

        Builder rotation(int rotation);

        SculkSensorPhase sculkSensorPhase();

        Builder sculkSensorPhase(SculkSensorPhase sculkSensorPhase);

        Shape shape();

        Builder shape(Shape shape);

        boolean short_();

        Builder short_(boolean short_);

        boolean signalFire();

        Builder signalFire(boolean signalFire);

        boolean slotZeroOccupied();

        Builder slotZeroOccupied(boolean slotZeroOccupied);

        boolean slotOneOccupied();

        Builder slotOneOccupied(boolean slotOneOccupied);

        boolean slotTwoOccupied();

        Builder slotTwoOccupied(boolean slotTwoOccupied);

        boolean slotThreeOccupied();

        Builder slotThreeOccupied(boolean slotThreeOccupied);

        boolean slotFourOccupied();

        Builder slotFourOccupied(boolean slotFourOccupied);

        boolean slotFiveOccupied();

        Builder slotFiveOccupied(boolean slotFiveOccupied);

        boolean snowy();

        Builder snowy(boolean snowy);

        int stage();

        Builder stage(int stage);

        South south();

        Builder south(South south);

        Thickness thickness();

        Builder thickness(Thickness thickness);

        Tilt tilt();

        Builder tilt(Tilt tilt);

        boolean triggered();

        Builder triggered(boolean triggered);

        Type typeData();

        Builder typeData(Type type);

        boolean unstable();

        Builder unstable(boolean unstable);

        boolean up();

        Builder up(boolean up);

        VerticalDirection verticalDirection();

        Builder verticalDirection(VerticalDirection verticalDirection);

        boolean waterlogged();

        Builder waterlogged(boolean waterlogged);

        East east();

        Builder east(East east);

        West west();

        Builder west(West west);

        Bloom bloom();

        Builder bloom(Bloom bloom);

        boolean cracked();

        Builder cracked(boolean cracked);

        boolean crafting();

        Builder crafting(boolean crafting);

        TrialSpawnerState trialSpawnerState();

        Builder trialSpawnerState(TrialSpawnerState trialSpawnerState);

        WrappedBlockState build();
    }
}

File: surf-api-core/surf-api-core/src/main/kotlin/dev/slne/surf/api/core/util/blockstate/BlockStateFactory.kt
package dev.slne.surf.api.core.util.blockstate

import com.github.retrooper.packetevents.protocol.world.BlockFace
import com.github.retrooper.packetevents.protocol.world.states.WrappedBlockState
import com.github.retrooper.packetevents.protocol.world.states.enums.*
import com.github.retrooper.packetevents.protocol.world.states.type.StateType
import kotlin.contracts.ExperimentalContracts
import kotlin.contracts.InvocationKind
import kotlin.contracts.contract

@Deprecated("Not longer maintained.")
interface BlockStateFactory {
    @Suppress("unused")
    interface Builder {
        fun age(): Int
        fun age(age: Int): Builder

        fun attached(): Boolean
        fun attached(attached: Boolean): Builder

        fun attachment(): Attachment
        fun attachment(attachment: Attachment): Builder

        fun axis(): Axis
        fun axis(axis: Axis): Builder

        fun berries(): Boolean
        fun berries(berries: Boolean): Builder

        fun bites(): Int
        fun bites(bites: Int): Builder

        fun bottom(): Boolean
        fun bottom(bottom: Boolean): Builder

        fun candles(): Int
        fun candles(candles: Int): Builder

        fun charges(): Int
        fun charges(charges: Int): Builder

        fun conditional(): Boolean
        fun conditional(conditional: Boolean): Builder

        fun delay(): Int
        fun delay(delay: Int): Builder

        fun disarmed(): Boolean
        fun disarmed(disarmed: Boolean): Builder

        fun distance(): Int
        fun distance(distance: Int): Builder

        fun down(): Boolean
        fun down(down: Boolean): Builder

        fun drag(): Boolean
        fun drag(drag: Boolean): Builder

        fun eggs(): Int
        fun eggs(eggs: Int): Builder

        fun enabled(): Boolean
        fun enabled(enabled: Boolean): Builder

        fun extended(): Boolean
        fun extended(extended: Boolean): Builder

        fun eye(): Boolean
        fun eye(eye: Boolean): Builder

        fun face(): Face
        fun face(face: Face): Builder

        fun facing(): BlockFace
        fun facing(facing: BlockFace): Builder

        fun half(): Half
        fun half(half: Half): Builder

        fun hanging(): Boolean
        fun hanging(hanging: Boolean): Builder

        fun hasBook(): Boolean
        fun hasBook(hasBook: Boolean): Builder

        fun hasBottle0(): Boolean
        fun hasBottle0(hasBottle0: Boolean): Builder

        fun hasBottle1(): Boolean
        fun hasBottle1(hasBottle1: Boolean): Builder

        fun hasBottle2(): Boolean
        fun hasBottle2(hasBottle2: Boolean): Builder

        fun hasRecord(): Boolean
        fun hasRecord(hasRecord: Boolean): Builder

        fun hatch(): Int
        fun hatch(hatch: Int): Builder

        fun hinge(): Hinge
        fun hinge(hinge: Hinge): Builder

        fun honeyLevel(): Int
        fun honeyLevel(honeyLevel: Int): Builder

        fun inWall(): Boolean
        fun inWall(inWall: Boolean): Builder

        fun instrument(): Instrument
        fun instrument(instrument: Instrument): Builder

        fun inverted(): Boolean
        fun inverted(inverted: Boolean): Builder

        fun layers(): Int
        fun layers(layers: Int): Builder

        fun leaves(): Leaves
        fun leaves(leaves: Leaves): Builder

        fun level(): Int
        fun level(level: Int): Builder

        fun lit(): Boolean
        fun lit(lit: Boolean): Builder

        fun locked(): Boolean
        fun locked(locked: Boolean): Builder

        fun mode(): Mode
        fun mode(mode: Mode): Builder

        fun moisture(): Int
        fun moisture(moisture: Int): Builder

        fun north(): North
        fun north(north: North): Builder

        fun note(): Int
        fun note(note: Int): Builder

        fun occupied(): Boolean
        fun occupied(occupied: Boolean): Builder

        fun shrieking(): Boolean
        fun shrieking(shrieking: Boolean): Builder

        fun canSummon(): Boolean
        fun canSummon(canSummon: Boolean): Builder

        fun open(): Boolean
        fun open(open: Boolean): Builder

        fun orientation(): Orientation
        fun orientation(orientation: Orientation): Builder

        fun part(): Part
        fun part(part: Part): Builder

        fun persistent(): Boolean
        fun persistent(persistent: Boolean): Builder

        fun pickles(): Int
        fun pickles(pickles: Int): Builder

        fun power(): Int
        fun power(power: Int): Builder

        fun powered(): Boolean
        fun powered(powered: Boolean): Builder

        fun rotation(): Int
        fun rotation(rotation: Int): Builder

        fun sculkSensorPhase(): SculkSensorPhase
        fun sculkSensorPhase(sculkSensorPhase: SculkSensorPhase): Builder

        fun shape(): Shape
        fun shape(shape: Shape?): Builder

        fun short(): Boolean
        fun short(short: Boolean): Builder

        fun signalFire(): Boolean
        fun signalFire(signalFire: Boolean): Builder

        fun slotZeroOccupied(): Boolean
        fun slotZeroOccupied(slotZeroOccupied: Boolean): Builder

        fun slotOneOccupied(): Boolean
        fun slotOneOccupied(slotOneOccupied: Boolean): Builder

        fun slotTwoOccupied(): Boolean
        fun slotTwoOccupied(slotTwoOccupied: Boolean): Builder

        fun slotThreeOccupied(): Boolean
        fun slotThreeOccupied(slotThreeOccupied: Boolean): Builder

        fun slotFourOccupied(): Boolean
        fun slotFourOccupied(slotFourOccupied: Boolean): Builder

        fun slotFiveOccupied(): Boolean
        fun slotFiveOccupied(slotFiveOccupied: Boolean): Builder

        fun snowy(): Boolean
        fun snowy(snowy: Boolean): Builder

        fun stage(): Int
        fun stage(stage: Int): Builder

        fun south(): South
        fun south(south: South): Builder

        fun thickness(): Thickness
        fun thickness(thickness: Thickness): Builder

        fun tilt(): Tilt
        fun tilt(tilt: Tilt): Builder

        fun triggered(): Boolean
        fun triggered(triggered: Boolean): Builder

        fun typeData(): Type
        fun typeData(type: Type): Builder

        fun unstable(): Boolean
        fun unstable(unstable: Boolean): Builder

        fun up(): Boolean
        fun up(up: Boolean): Builder

        fun verticalDirection(): VerticalDirection
        fun verticalDirection(verticalDirection: VerticalDirection): Builder

        fun waterlogged(): Boolean
        fun waterlogged(waterlogged: Boolean): Builder

        fun east(): East
        fun east(east: East): Builder

        fun west(): West
        fun west(west: West): Builder

        fun bloom(): Bloom
        fun bloom(bloom: Bloom): Builder

        fun cracked(): Boolean
        fun cracked(cracked: Boolean): Builder

        fun crafting(): Boolean
        fun crafting(crafting: Boolean): Builder

        fun trialSpawnerState(): TrialSpawnerState
        fun trialSpawnerState(trialSpawnerState: TrialSpawnerState): Builder

        fun build(): WrappedBlockState
    }

    @OptIn(ExperimentalContracts::class)
    companion object {
        @Suppress("DEPRECATION")
        @JvmStatic
        fun builder(stateType: StateType): Builder =
            BlockStateFactoryImpl.BuilderImpl(stateType.createBlockState().clone())

        @Suppress("DEPRECATION")
        @JvmStatic
        fun builder(wrappedBlockState: WrappedBlockState): Builder =
            BlockStateFactoryImpl.BuilderImpl(wrappedBlockState.clone())

        fun build(stateType: StateType, block: Builder.() -> Unit): WrappedBlockState {
            contract {
                callsInPlace(block, InvocationKind.EXACTLY_ONCE)
            }

            return builder(stateType).apply(block).build()
        }

        fun build(
            wrappedBlockState: WrappedBlockState,
            block: Builder.() -> Unit
        ): WrappedBlockState {
            contract {
                callsInPlace(block, InvocationKind.EXACTLY_ONCE)
            }

            return builder(wrappedBlockState).apply(block).build()
        }

        @JvmStatic
        fun of(stateType: StateType): WrappedBlockState = stateType.createBlockState().clone()
    }
}
